File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.py
class Trie:
   def __init__(self):
       self.node = {}      # 자식 노드들을 저장할 딕셔너리
       self.is_end = False # 단어의 끝 표시

   def insert(self, word: str) -> None:
       curr = self         # 루트부터 시작
       for c in word:      # 각 문자 순회
           if c not in curr.node:  # 문자가 없으면 새 트라이 생성
               curr.node[c] = Trie()
           curr = curr.node[c]     # 다음 노드로 이동
       curr.is_end = True  # 단어 끝 표시

   def search(self, word: str) -> bool:
       curr = self
       for c in word:      # 각 문자 존재 확인
           if c not in curr.node:
               return False
           curr = curr.node[c]
       return curr.is_end  # 정확한 단어면 True

   def startsWith(self, prefix: str) -> bool:
       curr = self
       for c in prefix:    # prefix의 모든 문자 존재 확인
           if c not in curr.node:
               return False
           curr = curr.node[c]
       return True         # 경로만 존재하면 True

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.cpp
class Trie {
public:
    Trie() {
        
    }
    
    void insert(string word) {
        Trie* curr = this;
        for(char c : word) {
            if (!curr->node.contains(c)) {
                curr->node[c] = new Trie();
            }
            curr = curr->node[c];
        }
        curr->is_end = true;
    }
    
    bool search(string word) {
        Trie* curr = this;
        for (char c : word) {
            if (!curr->node.contains(c)){
                return false;
            }
            curr = curr->node[c];
        }
        return curr->is_end;
    }
    
    bool startsWith(string prefix) {
        Trie* curr = this;
        for (char c : prefix) {
            if (!curr->node.contains(c)){
                return false;
            }
            curr = curr->node[c];
        }
        return true;
    }

private:
    unordered_map<char, Trie*> node;
    bool is_end = false;
};
/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */